helix-core/src/line_ending.rs

```diff
@@ -0,0 +1,121 @@
+use crate::{Rope, RopeGraphemes, RopeSlice};
+
+/// Represents one of the valid Unicode line endings.
+#[derive(PartialEq, Copy, Clone, Debug)]
+pub enum LineEnding {
+    Crlf, // CarriageReturn followed by LineFeed
+    LF,   // U+000A -- LineFeed
+    CR,   // U+000D -- CarriageReturn
+    Nel,  // U+0085 -- NextLine
+    LS,   // U+2028 -- Line Separator
+    VT,   // U+000B -- VerticalTab
+    FF,   // U+000C -- FormFeed
+    PS,   // U+2029 -- ParagraphSeparator
+}
+
+pub fn rope_slice_to_line_ending(g: &RopeSlice) -> Option<LineEnding> {
+    if let Some(text) = g.as_str() {
+        str_to_line_ending(text)
+    } else if g == "\u{000D}\u{000A}" {
+        Some(LineEnding::Crlf)
+    } else {
+        // Not a line ending
+        None
+    }
+}
+
+pub fn str_to_line_ending(g: &str) -> Option<LineEnding> {
+    match g {
+        "\u{000D}\u{000A}" => Some(LineEnding::Crlf),
+        "\u{000A}" => Some(LineEnding::LF),
+        "\u{000D}" => Some(LineEnding::CR),
+        "\u{0085}" => Some(LineEnding::Nel),
+        "\u{2028}" => Some(LineEnding::LS),
+        // Not a line ending
+        _ => None,
+    }
+}
+
+pub fn auto_detect_line_ending(doc: &Rope) -> Option<LineEnding> {
+    // based on https://github.com/cessen/led/blob/27572c8838a1c664ee378a19358604063881cc1d/src/editor/mod.rs#L88-L162
+
+    let mut ending = None;
+    // return first matched line ending. Not all possible line endings are being matched, as they might be special-use only
+    for line in doc.lines().take(100) {
+        ending = match line.len_chars() {
+            1 => {
+                let g = RopeGraphemes::new(line.slice((line.len_chars() - 1)..))
+                    .last()
+                    .unwrap();
+                rope_slice_to_line_ending(&g)
+            }
+            n if n > 1 => {
+                let g = RopeGraphemes::new(line.slice((line.len_chars() - 2)..))
+                    .last()
+                    .unwrap();
+                rope_slice_to_line_ending(&g)
+            }
+            _ => None,
+        };
```

Comment:
This is originally my code, so I'm really just commenting on my past self here, ha ha.  I'm looking at it now and I'm realizing this is wrong.  We need to handle the 1 vs 2-character grapheme cases in a different way.

I think what we actually need can be accomplished pretty nicely by taking advantage of some of `Option`s chaining methods:
```rust
// Last character as str.
let g1 = line.slice(line.len_chars().saturating_sub(1)..).as_str().unwrap();

// Last two characters as str, or empty str if they're not contiguous.
// It's fine to punt on the non-contiguous case, because Ropey guarantees
// that CRLF is always contiguous.
let g2 = line.slice(line.len_chars().saturating_sub(2)..).as_str().unwrap_or("");

// First check the two-character case for CRLF, then check the single-character case.
ending = str_to_line_ending(g2).or_else(|_| str_to_line_ending(g1));
```

You can in-line `g1` and `g2` if you want.

I think this is also a valid implementation for `rope_slice_to_line_ending()` (which I'm not totally sure is currently correct), so you put this code in there instead, and just call that here.